flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/BenchmarkUtils.java

```diff
@@ -0,0 +1,151 @@
+                ReadWriteUtils.instantiateWithParams((Map<String, ?>) params.get("inputs"));
+        DataGenerator modelDataGenerator = null;
+        if (params.containsKey("modelData")) {
+            modelDataGenerator =
+                    ReadWriteUtils.instantiateWithParams((Map<String, ?>) params.get("modelData"));
+        }
+
+        return runBenchmark(tEnv, name, stage, inputsGenerator, modelDataGenerator);
+    }
+
+    /**
+     * Executes a benchmark from a stage with its inputsGenerator in the provided environment.
+     *
+     * @return Results of the executed benchmark.
+     */
+    public static BenchmarkResult runBenchmark(
+            StreamTableEnvironment tEnv,
+            String name,
+            Stage<?> stage,
+            DataGenerator<?> inputsGenerator)
+            throws Exception {
+        return runBenchmark(tEnv, name, stage, inputsGenerator, null);
+    }
+
+    /**
+     * Executes a benchmark from a stage with its inputsGenerator and modelDataGenerator in the
+     * provided environment.
+     *
+     * @return Results of the executed benchmark.
+     */
+    public static BenchmarkResult runBenchmark(
+            StreamTableEnvironment tEnv,
+            String name,
+            Stage<?> stage,
+            DataGenerator<?> inputsGenerator,
+            DataGenerator<?> modelDataGenerator)
+            throws Exception {
+        StreamExecutionEnvironment env = TableUtils.getExecutionEnvironment(tEnv);
+
+        Table[] inputTables = inputsGenerator.getData(tEnv);
+        if (modelDataGenerator != null) {
+            ((Model<?>) stage).setModelData(modelDataGenerator.getData(tEnv));
+        }
+
+        Table[] outputTables;
+        if (stage instanceof Estimator) {
+            outputTables = ((Estimator<?, ?>) stage).fit(inputTables).getModelData();
+        } else if (stage instanceof AlgoOperator) {
+            outputTables = ((AlgoOperator<?>) stage).transform(inputTables);
+        } else {
+            throw new IllegalArgumentException("Unsupported Stage class " + stage.getClass());
+        }
+
+        for (Table table : outputTables) {
+            tEnv.toDataStream(table).addSink(new CountingAndDiscardingSink<>());
+        }
+
+        JobExecutionResult executionResult = env.execute();
+
+        BenchmarkResult result = new BenchmarkResult();
+        result.name = name;
+        result.totalTimeMs = (double) executionResult.getNetRuntime(TimeUnit.MILLISECONDS);
+        result.inputRecordNum = inputsGenerator.getNumData();
+        result.inputThroughput = result.inputRecordNum * 1000.0 / result.totalTimeMs;
+        result.outputRecordNum =
+                executionResult.getAccumulatorResult(CountingAndDiscardingSink.COUNTER_NAME);
+        result.outputThroughput = result.outputRecordNum * 1000.0 / result.totalTimeMs;
+
+        return result;
+    }
+
+    /** Prints out the provided benchmark result. */
+    public static void printResult(BenchmarkResult result) {
```

Comment:
I personally find that the output from `saveResultsAsJson(...)` to be more compact and readable than the strings printed by this method. Thus I am wondering if it is simpler to remove this method. What do you think?

Reply:
I think we still need to print out the benchmark results in the terminal. I'll rewrite `printResult` so that it prints in json format.